### DISCUSSION TITLE
QQL support for Double aggregation

### DISCUSSION THREAD

**Discussion by ShivaniSK** [finos/TimeBase-CE#236] [Q&A]
Hi,
We are trying to implement double aggregation in timebase query.
**Use-case:** 
Summation of bid size and ask sizes for levels 1,2 and 3 aggregated by timestamp, symbol.
Post that we want to average out the difference in the above summation of bid size and ask sizes by rolling up at minute level.

Below is our exact query which we are trying to execute

`select avg{} (BID_MINUS_ASK) from (select sum{} (bid_size) as BID_SIZE, sum{} (ask_size) AS ASK_SIZE, (BID_SIZE - ASK_SIZE) as BID_MINUS_ASK, SYMBOL  from BBO_stream where timestamp between '2022-08-25 10:00:00.000'd and '2022-08-25 11:00:00.000'd and symbol=='EURUSD' GROUP BY timestamp) over TIME (1m) `

However, its failing with Null pointer exception in timebase shell.

**Can you please suggest how to implement this as sub-queries are not supported?**

Regards,
Shivani

**Comment by alex-karpovich** [comment]
we do not support inner selects yet

I suggest to use next approach:
````
select AVG{} (bid_size - ask_size) as "average" FROM BBO_stream over TIME (1m) group by symbol
````
Description of this statement:          
for each minute will be calculated average of ```(bid_size - ask_size)``` for all messages matched criteria grouped by symbol

**Reply by ShivaniSK** [reply]
Hello @alex-karpovich ,
As per our use-case we have data like below

```
epochtime     bid_size  ask_size
11:01:22.200  12          10
**11:01:23.222  13          10
11:01:23.222  14          12**
11:02:20.122  11          11
11:03:25.800  11          11 
11:04:26.750  11          11
```

What we want is to first group similar epochtimes (11:01:23.222 in above example) and sum up the bid_size and ask_size for similar epochtimes. The result should be like below 

```
epochtime     bid_size  ask_size  ask_minus_bid
11:01:22.200  12          10              2
**11:01:23.222  27          22**        5
11:02:20.122  11          11              0
11:03:25.800  11          11              0 
11:04:26.750  11          11              0
```

Post that we want to apply a roll up over 1 minute which gives average of bid - ask

```
epochtime     average
11:01           3.5
11:02            0
11:03            0
11:04            0
```


Can you please suggest a way to acheive this without using inner query?

Regards,
Shivani

**Reply by alex-karpovich** [reply]
Greetings, @ShivaniSK 
For now I can suggest to use custom QQL functions - I will send you small guide next week.

Inner selects in our roadmap, but not soon.

**Reply by ShivaniSK** [reply]
Hello @alex-karpovich ,
can you please share the custom QQL guide as discussed.

Regards,
Shivani

**Comment by alex-karpovich** [comment]
Greetings,  please take a look here
https://github.com/epam/TimeBaseTutorial/tree/main/qql-functions

**Reply by ShivaniSK** [reply]
Hello @alex-karpovich , from what I understand the qql-functions will allow us to write custom aggregation functions right?
It still will not help us with use-case where "Inner query" is required?

**Comment by alex-karpovich** [comment]
@ShivaniSK we are going to support inner queries soon.    

For now, custom QQL functions may help you to process your data without inner queries

**Reply by ShivaniSK** [reply]
Sure @alex-karpovich will look more into QQL functions. Meanwhile is there any approximate timeline by which we can expect the inner query support?

**Reply by andymalakov** [reply]
Just a though - we typically implement custom bars like mentioned in your use case as "stream transformers". That is a piece of Java or Python code that reads source stream, computes custom aggregate function and writes it into output stream for subsequent pipelines.

Grouping market volume by timestamp and averaging will fit into ~10 lines of code.

**Reply by ShivaniSK** [reply]
Hi @alex-karpovich , can you please share any documentation or sample use-case implementation which uses stream transformation. That will be of great help.

**Reply by ShivaniSK** [reply]
Hi @alex-karpovich  can you please update?

**Reply by ShivaniSK** [reply]
Hi @alex-karpovich , do we have any documentation link or sample use-case implementation which uses stream transformation. That will be of great help.
